### PR TITLE
Python 3.12b2 testing

### PR DESCRIPTION
**Pull request by rwgk**
## Description
Enable **all** testing (test_embed) and `pip install setuptools` for `Setuptools helpers test` (setuptools is not longer installed by default: https://docs.python.org/3.12/whatsnew/3.12.html#removed).

First time all testing succeeds with Python 3.12!

## Suggested changelog entry:

```rst

```

### CODE REVIEW

**Comment by rwgk** [comment]
@henryiii

All unit tests (pytest & test_embed) work with Python 3.12.0b2, except the very last test in upstream.yml:

* https://github.com/pybind/pybind11/actions/runs/5228202706/jobs/9440398083

```
Setuptools helpers tests
...
tests/extra_setuptools/test_setuphelper.py FFFFFF                        [100%]
...
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '['/opt/hostedtoolcache/Python/3.12.0-beta.2/x64/bin/python', 'setup.py', 'build_ext', '--inplace']' returned non-zero exit status 1.
...

/opt/hostedtoolcache/Python/3.12.0-beta.2/x64/lib/python3.12/subprocess.py:571: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "/tmp/pytest-of-runner/pytest-0/test_simple_setup_py_11_False_0/setup.py", line 4, in <module>
    from setuptools import setup, Extension
ModuleNotFoundError: No module named 'setuptools'
```

**Comment by henryiii** [comment]
Setuptools is no longer installed by default in 3.12. It’s just another library like any other.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -242,8 +244,6 @@ jobs:
         python -m pip install -r tests/requirements.txt
 
     - name: Configure
-      env:
-        SETUPTOOLS_USE_DISTUTILS: stdlib
```
Comment (henryiii) [inline .github/workflows/ci.yml:246]: Testing this - this is invalid in 3.12, there is no stdlib distutils, and we shouldn't be using before that either. Just looking to see what fails.

**Comment by rwgk** [comment]
Ignoring two PyPy test_iostream flakes.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (henryiii) [reply .github/workflows/ci.yml:246]: FYI, I expect this was to build numpy before wheels were available.

**Pull request merged**